File: python/tests/test_combinatorics.py
"""
Tests for combinatorics module.
"""

import pytest
from ptef.combinatorics import (
    count_tokens_up_to_999,
    count_tokens_up_to_N,
    count_syllables_up_to_N,
    count_connectives_up_to_N
)


def test_count_tokens_up_to_999():
    """Test token counting for 1-999."""
    counts = count_tokens_up_to_999()
    
    # Check that basic tokens are present
    assert "um" in counts
    assert "dois" in counts
    assert "e" in counts
    assert "mil" in counts
    
    # Check that counts are positive
    for token, count in counts.items():
        assert count > 0
    
    # Check that "um" appears at least once
    assert counts["um"] >= 1


def test_count_tokens_up_to_N_small():
    """Test token counting for small N."""
    # Test N=10
    token_counts, connective_counts = count_tokens_up_to_N(10)
    
    # Check that we have some tokens
    assert len(token_counts) > 0
    assert len(connective_counts) > 0
    
    # Check that "um" appears once
    assert token_counts["um"] == 1
    
    # Check that "e" appears in connectives
    assert "e" in connective_counts


def test_count_tokens_up_to_N_medium():
    """Test token counting for medium N."""
    # Test N=100
    token_counts, connective_counts = count_tokens_up_to_N(100)
    
    # Check that we have more tokens than for N=10
    assert len(token_counts) > 0
    assert len(connective_counts) > 0
    
    # Check that "um" appears multiple times
    assert token_counts["um"] > 1


def test_count_tokens_up_to_N_large():
    """Test token counting for large N."""
    # Test N=1000
    token_counts, connective_counts = count_tokens_up_to_N(1000)
    
    # Check that we have tokens
    assert len(token_counts) > 0
    assert len(connective_counts) > 0
    
    # Check that "mil" appears
    assert "mil" in token_counts


def test_count_tokens_consistency():
    """Test consistency between direct and fast counting."""
    # For small N, fast counting should match direct counting
    for N in [10, 50, 100]:
        fast_counts, fast_conn = count_tokens_up_to_N(N)
        
        # Direct counting
        direct_counts = {}
        direct_conn = {}
        for n in range(1, N + 1):
            from ptef.grammar import text_number
            tokens = text_number(n)
            for token in tokens:
                direct_counts[token] = direct_counts.get(token, 0) + 1
                if token == "e":
                    direct_conn[token] = direct_conn.get(token, 0) + 1
        
        # Compare counts
        all_tokens = set(fast_counts.keys()) | set(direct_counts.keys())
        for token in all_tokens:
            fast_count = fast_counts.get(token, 0)
            direct_count = direct_counts.get(token, 0)
            assert fast_count == direct_count, f"Token '{token}' count mismatch: fast={fast_count}, direct={direct_count}"


def test_count_syllables_up_to_N():
    """Test syllable counting up to N."""
    # Test N=10
    syllables = count_syllables_up_to_N(10)
    assert syllables > 0
    
    # Test N=100
    syllables_100 = count_syllables_up_to_N(100)
    assert syllables_100 > syllables  # Should be more syllables for larger N


def test_count_connectives_up_to_N():
    """Test connective counting up to N."""
    # Test N=10
    connectives = count_connectives_up_to_N(10)
    assert connectives >= 0
    
    # Test N=100
    connectives_100 = count_connectives_up_to_N(100)
    assert connectives_100 >= connectives  # Should be more or equal connectives for larger N


def test_count_tokens_policy():
    """Test policy parameter."""
    # Currently only R1 is supported
    counts, _ = count_tokens_up_to_N(10, policy="R1")
    assert len(counts) > 0
    
    with pytest.raises(ValueError):
        count_tokens_up_to_N(10, policy="R2")


def test_count_tokens_edge_cases():
    """Test edge cases for token counting."""
    # Test N=0
    counts, conn = count_tokens_up_to_N(0)
    assert len(counts) == 0
    assert len(conn) == 0
    
    # Test N=1
    counts, conn = count_tokens_up_to_N(1)
    assert counts["um"] == 1
    assert len(conn) == 0  # No connectives for single number


def test_count_tokens_scaling():
    """Test that token counting scales properly."""
    # Test that counts increase with N
    counts_10, _ = count_tokens_up_to_N(10)
    counts_100, _ = count_tokens_up_to_N(100)
    
    # Total token count should increase
    total_10 = sum(counts_10.values())
    total_100 = sum(counts_100.values())
    assert total_100 > total_10

File: r/tests/testthat/test-combinatorics.R
# Tests for combinatorics module

test_that("count_tokens_up_to_n works for small N", {
  # Test N=10
  result <- count_tokens_up_to_n(10)
  
  # Check that we have some tokens
  expect_true(length(result$tokens) > 0)
  expect_true(length(result$connectives) > 0)
  
  # Check that "um" appears once
  expect_equal(result$tokens["um"], 1)
  
  # Check that "e" appears in connectives
  expect_true("e" %in% names(result$connectives))
})

test_that("count_tokens_up_to_n works for medium N", {
  # Test N=100
  result <- count_tokens_up_to_n(100)
  
  # Check that we have more tokens than for N=10
  expect_true(length(result$tokens) > 0)
  expect_true(length(result$connectives) > 0)
  
  # Check that "um" appears multiple times
  expect_true(result$tokens["um"] > 1)
})

test_that("count_tokens_up_to_n works for large N", {
  # Test N=1000
  result <- count_tokens_up_to_n(1000)
  
  # Check that we have tokens
  expect_true(length(result$tokens) > 0)
  expect_true(length(result$connectives) > 0)
  
  # Check that "mil" appears
  expect_true("mil" %in% names(result$tokens))
})

test_that("count_tokens_up_to_n is consistent", {
  # For small N, fast counting should match direct counting
  for (N in c(10, 50, 100)) {
    fast_result <- count_tokens_up_to_n(N)
    
    # Direct counting
    direct_counts <- integer(0)
    direct_conn <- integer(0)
    for (n in 1:N) {
      tokens <- text_number(n)
      for (token in tokens) {
        if (token %in% names(direct_counts)) {
          direct_counts[token] <- direct_counts[token] + 1
        } else {
          direct_counts[token] <- 1
        }
        if (token == "e") {
          if ("e" %in% names(direct_conn)) {
            direct_conn["e"] <- direct_conn["e"] + 1
          } else {
            direct_conn["e"] <- 1
          }
        }
      }
    }
    
    # Compare counts
    all_tokens <- unique(c(names(fast_result$tokens), names(direct_counts)))
    for (token in all_tokens) {
      fast_count <- fast_result$tokens[token]
      if (is.na(fast_count)) fast_count <- 0
      direct_count <- direct_counts[token]
      if (is.na(direct_count)) direct_count <- 0
      expect_equal(fast_count, direct_count, 
                   info = paste("Token '", token, "' count mismatch: fast=", fast_count, ", direct=", direct_count, sep = ""))
    }
  }
})

test_that("count_syllables_up_to_n works", {
  # Test N=10
  syllables <- count_syllables_up_to_n(10)
  expect_true(syllables > 0)
  
  # Test N=100
  syllables_100 <- count_syllables_up_to_n(100)
  expect_true(syllables_100 > syllables)  # Should be more syllables for larger N
})

test_that("count_connectives_up_to_n works", {
  # Test N=10
  connectives <- count_connectives_up_to_n(10)
  expect_true(connectives >= 0)
  
  # Test N=100
  connectives_100 <- count_connectives_up_to_n(100)
  expect_true(connectives_100 >= connectives)  # Should be more or equal connectives for larger N
})

test_that("count_tokens_up_to_n policy parameter works", {
  # Currently only R1 is supported
  result <- count_tokens_up_to_n(10, policy = "R1")
  expect_true(length(result$tokens) > 0)
  
  expect_error(count_tokens_up_to_n(10, policy = "R2"), "Only R1 policy is currently supported")
})

test_that("count_tokens_up_to_n handles edge cases", {
  # Test N=0
  result <- count_tokens_up_to_n(0)
  expect_equal(length(result$tokens), 0)
  expect_equal(length(result$connectives), 0)
  
  # Test N=1
  result <- count_tokens_up_to_n(1)
  expect_equal(result$tokens["um"], 1)
  expect_equal(length(result$connectives), 0)  # No connectives for single number
})

test_that("count_tokens_up_to_n scales properly", {
  # Test that counts increase with N
  result_10 <- count_tokens_up_to_n(10)
  result_100 <- count_tokens_up_to_n(100)
  
  # Total token count should increase
  total_10 <- sum(result_10$tokens)
  total_100 <- sum(result_100$tokens)
  expect_true(total_100 > total_10)
})
